pkg/datasource/data.go

```diff
@@ -0,0 +1,419 @@
+	}
+	ds.registerPool()
+	// TODO: add more checks / validation
+	return ds, nil
+}
+
+func (ds *dataSource) registerPool() {
+	ds.dPool.New = func() any {
+		d := &data{
+			Payload: make([][]byte, ds.payloadCount),
+		}
+		for i := range d.Payload {
+			d.Payload[i] = make([]byte, 0)
+		}
+		return d
+	}
+}
+
+func (ds *dataSource) Name() string {
+	return ds.name
+}
+
+func (ds *dataSource) Type() Type {
+	return ds.dType
+}
+
+func (ds *dataSource) NewData() Data {
+	return ds.dPool.Get().(Data)
+}
+
+func (ds *dataSource) ByteOrder() binary.ByteOrder {
+	return ds.byteOrder
+}
+
+func resolveNames(id uint32, fields []*field, parentOffset uint32) (string, error) {
+	if id >= uint32(len(fields)) {
+		return "", errors.New("invalid parent id")
+	}
+	out := ""
+	if FieldFlagHasParent.In(fields[id].Flags) {
+		p, err := resolveNames(fields[id].Parent-parentOffset, fields, parentOffset)
+		if err != nil {
+			return "", errors.New("parent not found")
+		}
+		out = p + "."
+	}
+	out += fields[id].Name
+	return out, nil
+}
+
+// AddStaticFields adds a statically sized container for fields to the payload and returns an accessor for the
+// container; if you want to access individual fields, get them from the DataSource directly
+func (ds *dataSource) AddStaticFields(size uint32, fields []Field) (FieldAccessor, error) {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	idx := ds.payloadCount
+
+	// temporary write to newFields to not write to ds.fields in case of errors
+	var newFields []*field
+
+	parentOffset := len(ds.fields)
+
+	for _, f := range fields {
+		if _, ok := ds.fieldMap[f.FieldName()]; ok {
+			return nil, fmt.Errorf("field %q already exists", f.FieldName())
+		}
+		nf := &field{
+			Name:         f.FieldName(),
+			Index:        uint32(len(ds.fields) + len(newFields)),
+			PayloadIndex: idx,
+			Flags:        FieldFlagStaticMember.Uint32(),
+		}
+		staticField, ok := f.(StaticField)
+		if !ok {
+			return nil, fmt.Errorf("field %q is not statically sized or does not implement StaticField", nf.Name)
+		}
+		nf.Size = staticField.FieldSize()
+		nf.Offs = staticField.FieldOffset()
+		if s, ok := f.(TypedField); ok {
+			nf.Kind = s.FieldType()
+		}
+		if tagger, ok := f.(TaggedField); ok {
+			nf.Tags = tagger.FieldTags()
+		}
+		if s, ok := f.(FlaggedField); ok {
+			nf.Flags |= uint32(s.FieldFlags())
+		}
+		if s, ok := f.(ParentedField); ok {
+			parent := s.FieldParent()
+			if parent >= 0 {
+				nf.Parent = uint32(parent + parentOffset)                          // TODO: validate?
+				nf.Flags |= FieldFlagHasParent.Uint32() | FieldFlagHidden.Uint32() // default to hide subfields
+			}
+		}
+
+		if nf.Offs+nf.Size > size {
+			return nil, fmt.Errorf("field %q exceeds size of container (offs %d, size %d, container size %d)", nf.Name, nf.Offs, nf.Size, size)
+		}
+		newFields = append(newFields, nf)
+	}
+
+	var err error
+	for i, f := range newFields {
+		f.FullName, err = resolveNames(uint32(i), newFields, uint32(parentOffset))
```

Comment:
There's something off with this. The field name already has full hierarchy names from

https://github.com/inspektor-gadget/inspektor-gadget/blob/8cf93c706f73cdc9f70ec3acbec7b3925870e4e2/pkg/operators/ebpf/struct.go#L181

So this is producing things like ` dst.dst.l3.dst.l3.addr.dst.l3.addr.v4`.

Reply:
Ah, `Field` from the ebpf operator was already handling full names - will fix that. The question is how we want to handle subfields and their additional in the metadata.

a) flattened hierarchy
```yaml
fieldname.subfieldname:
  annotations:
  ...
fieldname.othersubfieldname:
  annotations:
  ...
```

b) full hierarchy
```yaml
fieldname:
  subfields:
    subfieldname:
      annotations:
      ...
    othersubfieldname:
      annotations:
      ...
```

Reply:
I agree option b is better.

Reply:
The function itself has been fixed - the discussion around the metadata remains and hasn't been addressed, yet (and IMHO it's out of scope for now).